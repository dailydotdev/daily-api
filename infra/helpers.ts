import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
