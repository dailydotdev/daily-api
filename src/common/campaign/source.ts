import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
