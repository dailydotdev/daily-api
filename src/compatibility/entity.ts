import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
