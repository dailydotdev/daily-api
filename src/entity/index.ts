export * from './user';
export * from './AdvancedSettings';
export * from './Alerts';
export * from './Banner';
export * from './Bookmark';
export * from './BookmarkList';
export * from './DevCard';
export * from './Feed';
export * from './FeedAdvancedSettings';
export * from './FeedSource';
export * from './FeedTag';
export * from './Integration';
export * from './posts';
export * from './PostReport';
export * from './PostTag';
export * from './Settings';
export * from './Source';
export * from './SourceDisplay';
export * from './SourceFeed';
export * from './SourceMember';
export * from './SourceRequest';
export * from './TagSegment';
export * from './View';
export * from './Comment';
export * from './Keyword';
export * from './PostKeyword';
export * from './CommentMention';
export * from './ReputationEvent';
export * from './Submission';
export * from './notifications';
export * from './Feature';
export * from './Invite';
export * from './ContentImage';
export * from './MarketingCta';
export * from './SquadPublicRequest';
