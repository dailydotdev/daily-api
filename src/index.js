import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import KoaPinoLogger from 'koa-pino-logger';
import Router from 'koa-router';
import etag from 'koa-etag';


import config from './config';
// import compress from './middlewares/compress';
import errorHandler from './middlewares/errorHandler';
import logger from './logger';

import health from './routes/health';
import sources from './routes/sources';
import posts from './routes/posts';
import publications from './routes/publications';
import tweet from './routes/tweet';
import settings from './routes/settings';
import feeds from './routes/feeds';
import notifications from './routes/notifications';
import tags from './routes/tags';

const app = new Koa();

app.keys = [config.cookies.secret];

app.proxy = true;

// TODO: disabled due to performance issues
// app.use(compress());

app.use(bodyParser());
app.use(KoaPinoLogger({ logger, useLevel: 'debug' }));
app.use(errorHandler());
app.use(etag());

// Machine-to-machine authentication
app.use((ctx, next) => {
  if (ctx.request.get('authorization') === `Service ${config.accessSecret}`
    && ctx.request.get('user-id') && ctx.request.get('logged-in')) {
    // eslint-disable-next-line
    ctx.state = {
      user: {
        userId: ctx.request.get('user-id'),
      },
      service: true,
    };
  } else {
    delete ctx.request.headers['user-id'];
    delete ctx.request.headers['logged-in'];
  }
  return next();
});

const router = new Router({
  prefix: '/v1',
});

router.use(feeds.routes(), feeds.allowedMethods());
router.use(sources.routes(), sources.allowedMethods());
router.use(posts.routes(), posts.allowedMethods());
router.use(publications.routes(), publications.allowedMethods());
router.use(tweet.routes(), tweet.allowedMethods());
router.use(settings.routes(), settings.allowedMethods());
router.use(notifications.routes(), notifications.allowedMethods());
router.use(tags.routes(), tags.allowedMethods());

app.use(router.routes(), router.allowedMethods());
app.use(health.routes(), health.allowedMethods());

export default app;
