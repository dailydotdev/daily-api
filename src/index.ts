import 'reflect-metadata';
import fastify, {
  FastifyRequest,
  FastifyInstance,
  FastifyError,
} from 'fastify';
import fastifyRawBody from 'fastify-raw-body';
import helmet from '@fastify/helmet';
import cors from '@fastify/cors';
import mercurius, { MercuriusError } from 'mercurius';
import MercuriusGQLUpload from 'mercurius-upload';
import MercuriusCache from 'mercurius-cache';
import { NoSchemaIntrospectionCustomRule } from 'graphql';
// import fastifyWebsocket from '@fastify/websocket';

import './config';

import tracking from './tracking';
import auth from './auth';
import compatibility from './compatibility';
import routes from './routes';
import { Context } from './Context';
import { schema } from './graphql';
import createOrGetConnection from './db';
import { stringifyHealthCheck } from './common';
import { GraphQLError } from 'graphql';
import cookie, { FastifyCookieOptions } from '@fastify/cookie';
import { getSubscriptionSettings } from './subscription';
import { ioRedisPool } from './redis';
import { loadFeatures } from './growthbook';
import { runInRootSpan } from './telemetry/opentelemetry';
import { loggerConfig } from './logger';

type Mutable<Type> = {
  -readonly [Key in keyof Type]: Type[Key];
};

const userExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined => (ctx.userId ? `user:${ctx.userId}` : undefined);

const trackingExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined =>
  ctx.trackingId ? `tracking:${ctx.trackingId}` : undefined;

// readiness probe is set failureThreshold: 2, periodSeconds: 2 (4s) + small delay
const GRACEFUL_DELAY = 2 * 2 * 1000 + 5000;

export default async function app(
  contextFn?: (request: FastifyRequest) => Context,
): Promise<FastifyInstance> {
  let isTerminating = false;
  const isProd = process.env.NODE_ENV === 'production';
  const connection = await runInRootSpan(
    'createOrGetConnection',
    createOrGetConnection,
  );

  const app = fastify({
    logger: loggerConfig,
    disableRequestLogging: true,
    trustProxy: true,
  });

  app.log.info('loading features');
  await loadFeatures(app.log);

  const gracefulShutdown = () => {
    app.log.info('starting termination');
    isTerminating = true;
    setTimeout(async () => {
      await app.close();
      await connection.destroy();
      await ioRedisPool.end();
      process.exit();
    }, GRACEFUL_DELAY);
  };
  process.on('SIGINT', gracefulShutdown);
  process.on('SIGTERM', gracefulShutdown);

  app.register(helmet);
  app.register(cors, {
    origin: isProd ? /daily\.dev$/ : true,
    credentials: true,
  });
  app.register(cookie, {
    secret: process.env.COOKIES_KEY,
  }) as FastifyCookieOptions;
  app.register(auth, { secret: process.env.ACCESS_SECRET });
  app.register(tracking);
  app.register(fastifyRawBody, {
    field: 'rawBody',
    global: false,
    encoding: 'utf8',
    runFirst: true,
  });

  app.setErrorHandler((err, req, res) => {
    req.log.error({ err }, err.message);
    res.code(500).send({ statusCode: 500, error: 'Internal Server Error' });
  });

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    if (isTerminating) {
      res.status(500).send(stringifyHealthCheck({ status: 'terminating' }));
    } else {
      res.send(stringifyHealthCheck({ status: 'ok' }));
    }
  });

  app.get('/liveness', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  // app.register(fastifyWebsocket, {
  //   options: {
  //     maxPayload: 1048576,
  //     verifyClient: (info, next) => next(true),
  //   },
  // });

  app.register(MercuriusGQLUpload, {
    maxFileSize: 1024 * 1024 * 20,
    maxFiles: 1,
  });

  app.register(mercurius, {
    schema,
    context:
      contextFn ?? ((request): Context => new Context(request, connection)),
    queryDepth: 10,
    subscription: getSubscriptionSettings(connection),
    // Disable GraphQL introspection in production
    graphiql: !isProd,
    validationRules: isProd && [NoSchemaIntrospectionCustomRule],
    errorFormatter(execution, ctx) {
      if (execution.errors?.length > 0) {
        const flatErrors = execution.errors.flatMap<GraphQLError>((error) => {
          if (error.originalError.name === 'FastifyError') {
            return (
              (error.originalError as MercuriusError<GraphQLError>).errors ||
              error
            );
          }
          return error;
        });
        return {
          statusCode: 200,
          response: {
            data: execution.data,
            errors: flatErrors.map((error): GraphQLError => {
              const newError = error as Mutable<GraphQLError>;
              if (!error.originalError) {
                newError.extensions = {
                  code: 'GRAPHQL_VALIDATION_FAILED',
                };
              } else if (error.originalError?.name === 'EntityNotFoundError') {
                newError.message = 'Entity not found';
                newError.extensions = {
                  code: 'NOT_FOUND',
                };
              } else if (
                (error.originalError as FastifyError)?.code ===
                'MER_ERR_GQL_PERSISTED_QUERY_NOT_FOUND'
              ) {
                app.log.debug(
                  { body: ctx?.reply?.request?.body },
                  'unknown query',
                );
              } else if (!error.extensions?.code) {
                app.log.warn(
                  {
                    err: error.originalError,
                    body: ctx?.reply?.request?.body,
                  },
                  'unexpected graphql error',
                );
                newError.message = 'Unexpected error';
                newError.extensions = {
                  code: 'UNEXPECTED',
                };
              }
              if (isProd) {
                newError.originalError = undefined;
              }
              return newError;
            }),
          },
        };
      }
      return {
        statusCode: 200,
        response: execution,
      };
    },
  });

  if (isProd) {
    app.register(MercuriusCache, {
      ttl: 10,
      policy: {
        Query: {
          searchBookmarksSuggestions: {
            extendKey: userExtendKey,
          },
          searchBookmarks: {
            extendKey: userExtendKey,
          },
          // anonymousFeed: {
          //   extendKey: userExtendKey,
          // },
          // feed: {
          //   extendKey: userExtendKey,
          // },
          sourceFeed: {
            extendKey: userExtendKey,
          },
          tagFeed: {
            extendKey: userExtendKey,
          },
          keywordFeed: {
            extendKey: userExtendKey,
          },
          searchPostSuggestions: {
            extendKey: userExtendKey,
          },
          searchPosts: {
            extendKey: userExtendKey,
          },
          authorFeed: {
            extendKey: userExtendKey,
          },
          mostUpvotedFeed: {
            extendKey: userExtendKey,
          },
          mostDiscussedFeed: {
            extendKey: userExtendKey,
          },
          randomTrendingPosts: {
            extendKey: userExtendKey,
          },
          randomSimilarPostsByTags: {
            extendKey: userExtendKey,
          },
          randomDiscussedPosts: {
            extendKey: userExtendKey,
          },
          tagsCategories: true,
          advancedSettings: {
            extendKey: trackingExtendKey,
          },
          banner: true,
          postByUrl: {
            extendKey: userExtendKey,
          },
          post: {
            extendKey: userExtendKey,
          },
          postUpvotes: true,
          searchTags: true,
          userReadingRankHistory: true,
          userReadHistory: true,
        },
      },
    });
  }

  app.register(compatibility, { prefix: '/v1' });
  app.register(routes, { prefix: '/' });

  return app;
}
