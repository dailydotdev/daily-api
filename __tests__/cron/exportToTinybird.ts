import { DataSource } from 'typeorm';
import { saveFixtures } from '../helpers';
import createOrGetConnection from '../../src/db';
import { ArticlePost, PostTag, Source, User } from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import {
  postAuthorsFixture,
  postTagsFixture,
  sharedPostsFixture,
  tinybirdPostsFixture,
} from '../fixture/post';
import {
  PostsMetadataRepositoryDependency,
  PostsRepositoryDependency,
  PostsMetadataRepository,
  PostsRepository,
  TinybirdExportService,
  TinybirdPost,
} from '../../src/cron/exportToTinybird';
import {
  ITinybirdClient,
  TinybirdDatasourceMode,
  PostDatasourceResult,
  QueryResult,
  fetchfn,
  TinybirdClient,
} from '../../src/common/tinybird';
import { FastifyBaseLogger } from 'fastify';

let con: DataSource;

beforeAll(async () => {
  con = await createOrGetConnection();
});

beforeEach(async () => {
  jest.clearAllMocks();
  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, User, postAuthorsFixture);
  await saveFixtures(con, ArticlePost, tinybirdPostsFixture);
  await saveFixtures(con, ArticlePost, sharedPostsFixture);
  await saveFixtures(con, PostTag, postTagsFixture);
  await con
    .getRepository(User)
    .save({ id: '1', name: 'Ido', image: 'https://daily.dev/ido.jpg' });
  await con.getRepository(User).save({
    id: '2',
    name: 'Lee',
    image: 'https://daily.dev/lee.jpg',
  });
});

describe('PostsRepository', () => {
  it('should return posts to export to tinybird with specific properties', async () => {
    const now = new Date();
    const latest = new Date(now.getTime() - 10000);
    const repo = new PostsRepository(con);

    const posts = await repo.getForTinybirdExport(latest);
    posts.forEach((post) => {
      post.created_at = '';
      post.metadata_changed_at = '';
    });
    expect(posts).toHaveLength(7);
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'scout1',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p1',
      metadata_changed_at: '',
      post_private: 0,
      post_type: 'article',
      source_id: 'a',
      source_type: 'machine',
      tags_str: 'javascript,webdev',
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'author2',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p2',
      metadata_changed_at: '',
      post_private: 0,
      post_type: 'article',
      source_id: 'b',
      source_type: 'machine',
      tags_str: null,
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0.5,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'scout3',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p3',
      metadata_changed_at: '',
      post_private: 0,
      post_type: 'article',
      source_id: 'c',
      source_type: 'machine',
      tags_str: null,
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 1,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'scout4',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p4',
      metadata_changed_at: '',
      post_private: 0,
      post_type: 'article',
      source_id: 'a',
      source_type: 'machine',
      tags_str: 'backend,data,javascript',
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'author5',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p5',
      metadata_changed_at: '',
      post_private: 0,
      post_type: 'article',
      source_id: 'b',
      source_type: 'machine',
      tags_str: 'html,javascript',
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: 'scout6',
      banned: 0,
      content_curation: ['c1', 'c2'],
      created_at: '',
      creator_twitter: null,
      id: 'p6',
      metadata_changed_at: '',
      post_private: 1,
      post_type: 'article',
      source_id: 'p',
      source_type: 'machine',
      tags_str: null,
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0,
    });
    expect(posts).toContainEqual<TinybirdPost>({
      author_id: null,
      banned: 0,
      content_curation: [],
      created_at: '',
      creator_twitter: null,
      id: 'squadP1',
      metadata_changed_at: '',
      post_private: 1,
      post_type: 'article',
      source_id: 'squad',
      source_type: 'squad',
      tags_str: null,
      flags_json_str:
        '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
      language: 'en',
      is_ai_probability: 0,
    });
  });
});

describe('PostsMetadataRepository', () => {
  const dataSource = 'datasource_mock';

  it('append success', async () => {
    const tinybirdMock = {
      postToDatasource: async (
        datasource: string,
        mode: TinybirdDatasourceMode,
        csv: string,
      ): Promise<PostDatasourceResult> => {
        expect(datasource).toEqual(dataSource);
        expect(mode).toEqual(TinybirdDatasourceMode.APPEND);
        expect(csv).toMatchSnapshot();

        return {} as PostDatasourceResult;
      },
    } as ITinybirdClient;

    const postsMetadataRepository = new PostsMetadataRepository(
      tinybirdMock,
      dataSource,
    );

    const posts: TinybirdPost[] = [
      {
        id: 'id1',
        author_id: 'author_id1',
        created_at: 'created_at',
        metadata_changed_at: 'metadata_changed_at',
        creator_twitter: 'creator_twitter',
        source_id: 'source_id',
        tags_str: 'tags_str',
        post_type: 'post_type',
        post_private: 1,
        content_curation: ['content_curation'],
        source_type: 'source_type',
        banned: 0,
        flags_json_str:
          '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
        language: 'en',
        is_ai_probability: 0,
      },
      {
        id: 'id2',
        author_id: 'author_id2',
        created_at: 'created_at2',
        metadata_changed_at: 'metadata_changed_at2',
        creator_twitter: 'creator_twitter2',
        source_id: 'source_id2',
        tags_str: 'tags_str2',
        post_type: 'post_type2',
        post_private: 1,
        content_curation: ['content_curation1', 'content_curation2'],
        source_type: 'source_type2',
        banned: 0,
        flags_json_str:
          '{"visible": true, "showOnFeed": true, "sentAnalyticsReport": true}',
        language: 'en',
        is_ai_probability: 0.5,
      },
    ];

    await postsMetadataRepository.append(posts);
  });

  it('latest success', async () => {
    const time = '2023-07-27T17:11:23Z';
    const tinybirdClientMock = {
      query: async (): Promise<QueryResult<unknown>> => {
        return {
          rows: 1,
          data: [
            {
              latest: time,
            },
          ],
        } as QueryResult<unknown>;
      },
    } as unknown as ITinybirdClient;

    const postsMetadataRepository = new PostsMetadataRepository(
      tinybirdClientMock,
      dataSource,
    );

    const latest = await postsMetadataRepository.latest();
    expect(latest).toEqual(new Date(time));
  });

  it('latest tinybird returned zero rows', async () => {
    const tinybirdClientMock = {
      query: async (): Promise<QueryResult<unknown>> => {
        return {
          rows: 0,
        } as QueryResult<unknown>;
      },
    } as unknown as ITinybirdClient;

    const postsMetadataRepository = new PostsMetadataRepository(
      tinybirdClientMock,
      dataSource,
    );

    await expect(async () => {
      await postsMetadataRepository.latest();
    }).rejects.toThrow('no rows returned');
  });
});

describe('TinybirdExportService', () => {
  const mockLatest = new Date('2023-04-11 13:52:32.842861');

  it('should export correctly', async () => {
    const logger = {} as FastifyBaseLogger;
    const mockPosts = [
      {
        id: 'id',
      } as TinybirdPost,
    ];

    const postsRepositoryMock = {
      async getForTinybirdExport(latest: Date): Promise<TinybirdPost[]> {
        expect(latest).toEqual(mockLatest);
        return mockPosts;
      },
    } as PostsRepositoryDependency;

    const postsMetadataRepositoryMock = {
      async latest(): Promise<Date> {
        return mockLatest;
      },
      async append(posts: TinybirdPost[]): Promise<PostDatasourceResult> {
        expect(posts).toEqual(mockPosts);
        return {} as PostDatasourceResult;
      },
    } as PostsMetadataRepositoryDependency;

    const service = new TinybirdExportService(
      logger,
      postsRepositoryMock,
      postsMetadataRepositoryMock,
    );

    const result = await service.export();
    expect(result).toEqual({
      exported: 1,
      since: mockLatest,
      tinybird: {},
    });
  });

  it("shouldn't export if posts repository returned 0 posts", async () => {
    const logger = {} as FastifyBaseLogger;
    const postsRepositoryMock = {
      async getForTinybirdExport(): Promise<TinybirdPost[]> {
        return [];
      },
    } as PostsRepositoryDependency;

    const postsMetadataRepositoryMock = {
      async latest(): Promise<Date> {
        return mockLatest;
      },
    } as PostsMetadataRepositoryDependency;

    const service = new TinybirdExportService(
      logger,
      postsRepositoryMock,
      postsMetadataRepositoryMock,
    );

    const result = await service.export();
    expect(result).toEqual({
      exported: 0,
      since: mockLatest,
      tinybird: null,
    });
  });

  it('should log exception if occurred', async () => {
    const logger = {
      error: (obj, msg) => {
        expect(obj.err.message).toEqual('ooops!');
        expect(msg).toEqual('failed to replicate posts to tinybird');
      },
    } as FastifyBaseLogger;
    const postsRepositoryMock = {} as PostsRepositoryDependency;

    const postsMetadataRepositoryMock = {
      async latest(): Promise<Date> {
        throw new Error('ooops!');
      },
    } as PostsMetadataRepositoryDependency;

    const service = new TinybirdExportService(
      logger,
      postsRepositoryMock,
      postsMetadataRepositoryMock,
    );

    await service.exportAndLog();
  });

  it('should log if no posts to be exported', async () => {
    const logger = {
      info: (obj, msg) => {
        expect(obj).toEqual({
          exported: 0,
          since: mockLatest,
          tinybird: null,
        });
        expect(msg).toEqual('no posts to replicate');
      },
    } as FastifyBaseLogger;
    const postsRepositoryMock = {
      async getForTinybirdExport(): Promise<TinybirdPost[]> {
        return [];
      },
    } as PostsRepositoryDependency;
    const postsMetadataRepositoryMock = {
      async latest(): Promise<Date> {
        return mockLatest;
      },
    } as PostsMetadataRepositoryDependency;

    const service = new TinybirdExportService(
      logger,
      postsRepositoryMock,
      postsMetadataRepositoryMock,
    );

    await service.exportAndLog();
  });

  it('should log about posts successfully replicated', async () => {
    const logger = {
      info: (obj, msg) => {
        expect(obj).toEqual({
          exported: 1,
          since: mockLatest,
          tinybird: {},
        });
        expect(msg).toEqual('posts replicated successfully to tinybird');
      },
    } as FastifyBaseLogger;
    const postsRepositoryMock = {
      async getForTinybirdExport(): Promise<TinybirdPost[]> {
        return [{} as TinybirdPost];
      },
    } as PostsRepositoryDependency;
    const postsMetadataRepositoryMock = {
      async latest(): Promise<Date> {
        return mockLatest;
      },
      async append(): Promise<PostDatasourceResult> {
        return {} as PostDatasourceResult;
      },
    } as PostsMetadataRepositoryDependency;

    const service = new TinybirdExportService(
      logger,
      postsRepositoryMock,
      postsMetadataRepositoryMock,
    );

    await service.exportAndLog();
  });

  test.skip('test real tinybird', async () => {
    const host = 'https://api.tinybird.co';
    const token = '';
    const tinybirdClient = new TinybirdClient(
      token,
      host,
      fetch as unknown as fetchfn,
    );

    const postsMetadataRepository = new PostsMetadataRepository(
      tinybirdClient,
      'posts_metadata',
    );

    const postsRepository = new PostsRepository(con);

    const exportService = new TinybirdExportService(
      console as unknown as FastifyBaseLogger,
      postsRepository,
      postsMetadataRepository,
    );

    await exportService.exportAndLog();
  });
});
