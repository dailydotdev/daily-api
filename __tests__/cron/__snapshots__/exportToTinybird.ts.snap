// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsMetadataRepository append success 1`] = `
"id1,author_id1,created_at,metadata_changed_at,creator_twitter,source_id,tags_str,0,post_type,1,['content_curation'],source_type,\\"{\\"\\"visible\\"\\": true, \\"\\"showOnFeed\\"\\": true, \\"\\"sentAnalyticsReport\\"\\": true}\\",en,0
id2,author_id2,created_at2,metadata_changed_at2,creator_twitter2,source_id2,tags_str2,0,post_type2,1,\\"['content_curation1', 'content_curation2']\\",source_type2,\\"{\\"\\"visible\\"\\": true, \\"\\"showOnFeed\\"\\": true, \\"\\"sentAnalyticsReport\\"\\": true}\\",en,0.5"
`;
